src/main/scala/dsls/regex/RegularExpression.scala

```diff
@@ -10,9 +13,50 @@ package dsls.regex
 /** The top of a class hierarchy that encodes regular expressions. */
 abstract class RegularExpression {
   /** returns true if the given string matches this regular expression */
+  
   def matches(string: String) = RegexMatcher.matches(string, this)
+  
+  // || represents the Union of this, which is of type RegularExpression, 
+  // with "other", which is also of type RegularExpression
+  def || (other: RegularExpression) = Union(this, other)
+  
+  // || represents the Concat of this, which is of type RegularExpression, 
+  // with "other", which is also of type RegularExpression
+  def ~ (other: RegularExpression) = Concat(this, other) 
+  
+  def <*> = Star (this)
+  
+  // another way to do it is def <+> = this ~ (this <*>)
+  def <+> = Concat (this, Star(this)) 
+  
+
+  def apply(n: Int): RegularExpression = {
+    n match {
+        case 0 => EPSILON // base case 
+        case n if n >= 1 => this ~ apply(n-1) // concat "this" with what "apply" recursively returns 
```

Comment:
If you are going to use the syntactic sugar version of the langugae here (i.e. `~` over, `Concat(_,_)`) then I would use it above as well when defining the other functions. Functionally, it doesn't matter, but I think it would be more consistent style.